### ISSUE TITLE
Stylesheets: Dark theme 2D geometry with little contrast to the background color

### ISSUE THREAD

**Issue by semhustej** [FreeCAD/FreeCAD#17979]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

There is too little contrast of 2D geometry with default color to the background color when using the default Dark Theme. It is not critical, it is possible to work in 2D with the default setup, but not too pleasant.

See screenshot of a 2D drawing:
![dark_mode_2D_geometry](https://github.com/user-attachments/assets/6181b49e-5b5a-4b09-8703-39241bff314e)

### Full version info

```shell
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.0.0RC4.39104 (Git) Conda
Build type: Release
Branch: (HEAD detached at 1.0rc4)
Hash: 83646753e051eaf438c0d5e9351cb57bd27377ab
Python 3.11.10, Qt 5.15.15, Coin 4.0.3, Vtk 9.3.0, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
```

### Subproject(s) affected?

Core

### Anything else?

I am not sure if there is a solution that would not break other things, especially rendering of 3D objects.

In standard 2D drafting software, the default black lines are usually rendered white if user chooses a black background.

I know it is possible to adjust the background color, but a new user opening FreeCAD for the first time needs to ne taken into account, this issue is about the defaults.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by pierreporte** [comment]
Maybe it is possible to follow the approach for Sketcher when @kadet1090 fixed the problem for that workbench.

**Comment by maxwxyz** [comment]
@MisterMakerNL FYI

**Comment by MisterMakerNL** [comment]
This is one of the biggest bug in the UI atm.
I made a macro that fixes this so it is fixable... I think. 
The macro checks if there are faces linked to the object if not it applies the "_sketcher create line color_". 
If there are faces linked to the line or point it applies the "_default shape line color_". 

Now we have to implement this into Freecad. Not sure if we want too run this macro every-time the user creates an object in Part and in Draft? It seems not to be necessary in Partdesign. 

Could use some expert advice here @maxwxyz 

<img width="1651" height="1553" alt="Image" src="https://github.com/user-attachments/assets/832be545-40ef-431f-8132-a0f1a89beb66" />

> 
> import FreeCAD as App
> 
> def color_int_to_rgb_tuple(color_int):
>     # Convert integer to 8-digit hex string
>     hexstr = format(int(color_int), '08x')
>     # Parse as RGBA
>     r = int(hexstr[:2], 16)
>     g = int(hexstr[2:4], 16)
>     b = int(hexstr[4:6], 16)
>     a = int(hexstr[6:8], 16)  # Alpha, usually unused for colors in FreeCAD
>     print(f"Color int: {color_int} → R,G,B,A: {r},{g},{b},{a}")
>     return (r/255.0, g/255.0, b/255.0)
> 
> print("--- Fetching colors from preferences (as RGBA) ---")
> param = App.ParamGet("User parameter:BaseApp/Preferences/View")
> part_line_color = color_int_to_rgb_tuple(param.GetUnsigned("DefaultShapeLineColor"))
> sketcher_line_color = color_int_to_rgb_tuple(param.GetUnsigned("CreateLineColor"))
> print(f"Part/PartDesign line color: {part_line_color}")
> print(f"Sketcher create line color: {sketcher_line_color}")
> 
> doc = App.ActiveDocument
> colored_something = False
> 
> if doc is None:
>     print("No active document found.")
> else:
>     print("Iterating over document objects...")
>     for obj in doc.Objects:
>         print(f"Checking object: {obj.Name}")
>         if hasattr(obj, "Shape") and hasattr(obj, "ViewObject") and obj.ViewObject.Visibility:
>             edge_count = len(obj.Shape.Edges)
>             face_count = len(obj.Shape.Faces)
>             vertex_count = len(obj.Shape.Vertexes)
>             print(f"  Edges: {edge_count}, Faces: {face_count}, Vertexes: {vertex_count}")
>             # For objects with faces → use DefaultShapeLineColor
>             if face_count > 0:
>                 if edge_count > 0:
>                     obj.ViewObject.LineColor = part_line_color
>                     print(f"  Set LineColor to {part_line_color}")
>                 if vertex_count > 0 and hasattr(obj.ViewObject, "PointColor"):
>                     obj.ViewObject.PointColor = part_line_color
>                     print(f"  Set PointColor to {part_line_color}")
>                 colored_something = True
>             # For objects without faces → use CreateLineColor
>             elif edge_count > 0:
>                 obj.ViewObject.LineColor = sketcher_line_color
>                 print(f"  Set LineColor to {sketcher_line_color}")
>                 if vertex_count > 0 and hasattr(obj.ViewObject, "PointColor"):
>                     obj.ViewObject.PointColor = sketcher_line_color
>                     print(f"  Set PointColor to {sketcher_line_color}")
>                 colored_something = True
>             # For point-only objects (no edges, no faces)
>             elif vertex_count > 0:
>                 if hasattr(obj.ViewObject, "PointColor"):
>                     obj.ViewObject.PointColor = sketcher_line_color
>                     print(f"  Set PointColor to {sketcher_line_color}")
>                     colored_something = True
>         else:
>             print(f"  Object {obj.Name} skipped (no Shape or not visible)")
> 
> if colored_something:
>     print("Colored at least one object.")
>     App.Console.PrintMessage("Colored all objects: DefaultShapeLineColor=with faces, CreateLineColor=without faces or points only.\n")
> else:
>     print("No objects with edges or points found to color.")
>     App.Console.PrintError("No objects with edges or points found.\n")

[Edge_color.zip](https://github.com/user-attachments/files/22026230/Edge_color.zip)

**Comment by maxwxyz** [comment]
I don't think this is a good idea.
The goal is to separate shape colors (material appearance) and theme appearance of the shapes in the app. The latter should be defined by theme. Other CADs have an additional Draw Style for viewing the Material appearance which is used for exporting the models.
@kadet1090 FYI

**Comment by MisterMakerNL** [comment]
I agree with that, but I think 2D geometry should be different than 3D geometry and this code can do this.
But I also was discussing this with Kadet on discord, but I kinda came too the conclusion that nobody seems to care. Like this bug makes it pretty much impossible to use draft with dark theme. And I can think that existing users know how to work around it. But new users would definitely come here to complain about this.

So this issue might not be a big of an issue as thought, and Draft wb is pretty much a dead wb used to compensate for missing functions in other wb.

TLDR nobody cares.

**Comment by maxwxyz** [comment]
There is no difference between 2D or 3D shapes, all is a shape here. And Draft WB is not dead.

**Comment by semhustej** [comment]
@MisterMakerNL thanks for looking into the issue.

I think the proposed solution with macro changing line and point colors based on whether the object has face / is faceless is not a good solution from the Draft and BIM workbench user perspective.

These are the reasons I see:

- in Draft/BIM the line/point color might not there just to represent the object in the 3D view. It can have a meaning and data associated with the color. It is possible to set different color to 3D view when the line is on a TechDraw page, but that is not a solution.
- this solution brings inconsistency - some lines are black, some white with no user input
- this solution will confuse users, as objects might be 'created' with different color than the one set as a current color
- this solution will confuse users, as colors of objects might change unexpectedly (when user changes the `Make Face` property, the color of the object lines changes)
- this solution disables users to use other line colors than white/black. I am sure this could be fixed though.


I agree with @maxwxyz here:
>
>The goal is to separate shape colors (material appearance) and theme appearance of the shapes in the app. The latter should be defined by theme. Other CADs have an additional Draw Style for viewing the Material appearance which is used for exporting the models


The standard solution in AutoCAD based drafting tools is to use one color (it is index color nr. 7) as a special color. This color

- is printed/exported always black
- if a background is set to a dark color, objects using the nr.7 color are displayed white in the 3D view
- if a background is set to a light color, objects using the nr.7 color are displayed black in the 3D view


Note there has already been discussion about this on the forum, apparently without a conclusion:
https://forum.freecad.org/viewtopic.php?style=3&t=89474

**Comment by MisterMakerNL** [comment]
Can you show some examples what you do with draft? What is it used for?

**Comment by semhustej** [comment]
An example of pure Draft WB output:
https://forum.freecad.org/viewtopic.php?t=99093

Note that Draft workbench is a key workbench for creating construction documentation from the BIM workbench models, as described in this tutorial (the Shape Views method): https://wiki.freecad.org/Arch_tutorial#2D_drawings
This is used for example in the BIMexample file which is part of FreeCAD installation:

![Image](https://github.com/user-attachments/assets/ab2e3a50-8800-49c2-b7f4-57a93913d86e)


The AEC industry relies heavily on printed construction documentation. I did not go through all of them, but i think most of these examples would rely in one way or another on some Draft objects used to document the model: https://wiki.osarch.org/index.php?title=FreeCAD/Architecture_3D_models_created_in_FreeCAD

Draft workbench is very flexible. I think that everything can be drafted in the workbench. I am posting only AEC examples.

**Comment by MisterMakerNL** [comment]
Reading the forum and the last post I think this can only be fixed by giving draftsight object their own sketcher like color system. Which I think was @kadet1090 opinion too.

**Comment by semhustej** [comment]
> Reading the forum and the last post I think this can only be fixed by giving draftsight object their own sketcher like color system. Which I think was [@kadet1090](https://github.com/kadet1090) opinion too.

It seems so.

It is not an issue that needs to be solved for the 1.1 release, as it is there already from the 1.0 release. Draft users just have to switch to the light theme.

**Comment by DesignWeaver3D** [comment]
I feel like this issue is being downplayed too much. 3D Wireframe mode is useless with the darkness of the background. Please, reduce the default Dark Theme background darkness to support sufficient contrast with the default 3D edge color.

<img width="423" height="405" alt="Image" src="https://github.com/user-attachments/assets/e7073908-f3ae-486f-aeb6-b5b9e66b24d1" />

Can you see the wireframe? It's in there.

Here it is again while the body is selected.

<img width="457" height="406" alt="Image" src="https://github.com/user-attachments/assets/8742b7e6-2633-4cb3-b304-f3a65b794c67" />

EDIT: Sorry, I probably should have posted in this issue: https://github.com/FreeCAD/FreeCAD/issues/25738

**Comment by pierreporte** [comment]
This is not a theme issue, it’s a core issue. Sketches had the same problem until the default color was changed to a hardcoded value to a theme value.

**Comment by kevenwyld** [comment]
As a dark theme user this is quite important to me, and I also came to the same conclusion as above which is that wires only have one color throughout the application, but in reality are representing two different distinct use cases which require opposite colors to contrast with nearby geometry.

Case 1: When wires are edges of 3d geometry, for example a padded shape in part design, where line color should contrast with object color
Case 2: When wires are drawn alone on the 3d view background, where they should contrast with the background it self.

EDIT: Just to add that me and probably others use FreeCAD because it has a dark theme in addition to all the other benefits.  In contrast (hehe) to many industry cad/cam packages which have no dark theme at all and refuse to introduce one.

EDIT2: Video from previous issue:

https://github.com/user-attachments/assets/fce63580-6e9a-4866-adb1-e43ee28dae96

I'm not sure I'd agree that this is a feature request.  Seems more like dark mode where you can't see what you draw is actually a problem since dark mode appears to be core functionality (new FC installs prompt you to select a theme on first start).

**Comment by Roy-043** [comment]
It should be pointed out that an object with faces can have loose edges. So there are not 2 cases, but at least 3. Also note that the Part WB can create primitives without faces. So the suggestion to fix this with an additional Draft preference does not make sense.

IMO this is a theming issue, but even if it is not, adapting the dark theme is easy and would improve the current situation.

**Comment by Roy-043** [comment]
### Proposal

<img width="888" height="418" alt="Image" src="https://github.com/user-attachments/assets/37b0df22-7952-45ef-a57d-59bc1b6e0892" />

#### Same colors on a white background

<img width="888" height="418" alt="Image" src="https://github.com/user-attachments/assets/cbc7d9b7-efb4-42a7-8f6f-3f0a0b686f2b" />

#### Same colors on the 'classic' background

<img width="916" height="407" alt="Image" src="https://github.com/user-attachments/assets/03685cd1-fdd8-4843-8636-cfa66c4366e8" />

**Comment by maxwxyz** [comment]
To add: this is not a theme-only change but this changes the actual geometry. If the geometry is exported e.g. as STEP file, the edges will be light grey. If the file is opened in FreeCAD light, the edges are similar to the light background color, same if the STEP gets imported again.
IMO it would be better to change the almost black background in the dark theme to a lighter dark

**Comment by Roy-043** [comment]
It does not change geometry, but the colors are kept when switching themes. The proposed colors look OK on a white background, but not so good on the 'classic' background. See added images above.

**Comment by R2V1W3** [comment]
Draft lines need their own color setting parameter back, removing draft dedicated visual parameters was wrong decision IMO. 
I am sure people will settle even if the line color default uses draft annotation color parameter, or the sketcher line color if adding new parameters is a problem. Managing color through layers isn't working because many tools create draft objects before user has a chance to create suitable colored layer. 
The line color also applies to line color during creation of object, dragging line or polyline, as an example.  

Drafts are among the most important objects in many workflows.

Wireframe also needs its own color parameter, but that should be another issue it doesn't have to be solved at once if it means Draft will be waiting for a solution indefinitely.

**Comment by Roy-043** [comment]
@FreeCAD/design-working-group Maybe you can chip in here. For me this is primarily a theming issue. One way to solve it would be to change the background color in the Dark theme as Max has suggested (but that may be controversial 😄 )

**Comment by maxwxyz** [comment]
Workaround for now: https://github.com/FreeCAD/FreeCAD/pull/32827
